Replace executor with a thread safe implementation

jacobperron opened the pull request:
The rclcpp::MultiThreadedExecutor is not thread safe, and so it is possible to hit races that cause gazebo_ros to crash. Specifically, there's a race when adding and removing nodes, which can happen when node are created and destroyed during runtime.

This is a copy of rclcpp::Executor with some modifications to make it thread safe. A lock was added to the spin implementation and also to the add_node and remove_node methods. In order to avoid deadlock, we must have a timeout in the spin loop while waiting for work.

Fixes #1169

Replaces https://github.com/ros-simulation/gazebo_ros_pkgs/pull/1196

jacobperron committed:
Replace executor with a thread safe implementation

The rclcpp::MultiThreadedExecutor is not thread safe, and so it is possible to hit races that cause gazebo_ros to crash. Specifically, there's a race when adding and removing nodes, which can happen when node are created and destroyed during runtime.

This is a copy of rclcpp::Executor with some modifications to make it thread safe. A lock was added to the spin implementation and also to the add_node and remove_node methods. In order to avoid deadlock, we must have a timeout in the spin loop while waiting for work.

Fixes #1169

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron committed:
Fix lint

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron commented:
@osrf-jenkins run tests please

jacobperron commented:
@osrf-jenkins run tests please

scpeters commented:
this sounds reasonable to me, but perhaps someone with ros2 experience can review it?

scpeters commented:
@osrf-jenkins run tests please

scpeters commented:
I tested this in a rocker container, and `test_node` was failing every single time; it wasn't flaky at all. Running this branch, fixed it completely, so this looks like it fixes the failing test.

I'll defer to some else to review the actual code.

jacobperron commented:
Perhaps this is also a good opportunity to switch to a pimpl pattern for the executor to help maintain ABI stability going forward. I don't think we need to expose all of the API inherited from the rclcpp executor.

ivanpauno commented on gazebo_ros/include/gazebo_ros/executor.hpp:
```diff
@@ -34,18 +36,65 @@ class Executor : public rclcpp::executors::MultiThreadedExecutor
+   *
+   * \see rclcpp::Executor::add_node
+   */
+  virtual void
```
nit: this should be `override` and not `virtual`, same to the other methods here

jacobperron replied:
see ad69f15

ivanpauno commented on gazebo_ros/include/gazebo_ros/executor.hpp:
```diff
@@ -1,4 +1,4 @@
-// Copyright 2018 Open Source Robotics Foundation, Inc.
+// Copyright 2014 Open Source Robotics Foundation, Inc.
```
?

jacobperron replied:
Since this directly copied from [`rclcpp::MultiThreadedExecutor`](https://github.com/ros2/rclcpp/blob/adebda52c541533d878a4d0bae2ed112403c1e2a/rclcpp/include/rclcpp/executors/multi_threaded_executor.hpp), I kept the same copyright notice.

ivanpauno commented on gazebo_ros/src/executor.cpp:
```diff
@@ -34,14 +49,100 @@ Executor::~Executor()
+{
+  while (rclcpp::ok(this->context_) && spinning.load()) {
+    rclcpp::AnyExecutable any_exec;
+    std::lock_guard<std::mutex> wait_lock(wait_mutex_);
```
Why not just using the rclcpp `SingleThreadedExecutor`?

I mean, this is just creating N threads that are running this loop in a mutually exclusive fashion, so there's actually zero benefit of having many threads (the threads are all the time fighting for the lock, and only one is doing work).

jacobperron replied:
Yeah, I think you're right. I overlooked that when I increased the scope of the lock (compared to the upstream implementation in [`rclcpp::MultiThreadedExecutor`](https://github.com/ros2/rclcpp/blob/adebda52c541533d878a4d0bae2ed112403c1e2a/rclcpp/src/rclcpp/executors/multi_threaded_executor.cpp#L76-L114).

jacobperron replied:
I've restored the original scope of `wait_mutex_`, which I think keeps the multi-threaded nature of the executor. See f9c0b54.

ivanpauno commented on gazebo_ros/src/executor.cpp:
```diff
@@ -34,14 +49,100 @@ Executor::~Executor()
+      return;
+    }
+
+    if (!get_next_executable(any_exec, next_exec_timeout_)) {
```
why the timeout is needed?

I see you mentioned a deadlock in the PR post, but I'm not sure of how that's happening.

jacobperron replied:
I don't recall exactly, but I'mp pretty sure I was running into deadlock when adding/removing nodes from separate threads. I think with the changes in f9c0b54, a timeout is no longer required.

ivanpauno reviewed (changes requested):
I have left some comments.

I'm not convinced this is the right approach, but I haven't thought a good way of solving the issue (maybe we should be fixing the issue in `rclcpp`?).

jacobperron committed:
override instead of virtual

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron committed:
Add regression test for adding and removing nodes to executor

Repeatedly adding and removing nodes to an executor that is not thread-safe may cause a segfault.

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron committed:
Fix use of mutexes

Restore original scope for wait_mutex_.
Use a separate set of mutexes for making adding/removing nodes thread-safe.
The set of mutexes is used to give priority to threads doing the adding/removing so that they are not constantly blocked by the quick spin threads.

This eliminates the need for the timeout in the spin threads.

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron commented:
@ivanpauno Thanks for the review :bow: 

I've added a regression test (2c7dfa9) and made an improvement related to the use of mutexes (f9c0b54). Rather than reusing `wait_lock_` for adding/removing nodes, I've introduced a new mutex specifically for this case. Actually, I make use of three mutexes to give lock priority to threads calling `add_node()` and `remove_node()` (as described in [this SO post](https://stackoverflow.com/a/11673600))). I think the priority locking is necessary, otherwise calling `add_node()` and `remove_node()` may block for a long time while they wait on the spin threads. E.g. the regression test I've added runs significantly slower without the priority locking.

Using separate mutexes for adding/removing nodes also removes the need for the timeout in the spin threads.

> I'm not convinced this is the right approach, but I haven't thought a good way of solving the issue (maybe we should be fixing the issue in rclcpp?).

I agree that fixing this in rclcpp is the best option. Previously, I wasn't sure about the best way to fix the issue in a way that would be accepted in rclcpp, but perhaps with the recent changes I've pushed it would be accepted. What do you think?

ivanpauno commented on gazebo_ros/test/test_executor.cpp:
```diff
@@ -0,0 +1,50 @@
+#include <rclcpp/node.hpp>
+
+// Regression test for https://github.com/ros-simulation/gazebo_ros_pkgs/issues/1169
+TEST(TestExecutor, AddRemoveNodes)
```
shouldn't the test have another thread spinning?
If not, I don't think it can fail

jacobperron replied:
The `Executor` starts a spin thread on construction. I've confirmed that this test fails if I revert the changes to the executor.

ivanpauno replied:
ah, I forgot this extension works differently.

ivanpauno commented on gazebo_ros/include/gazebo_ros/executor.hpp:
```diff
@@ -87,8 +87,13 @@ class Executor : public rclcpp::Executor
   std::mutex wait_mutex_;
+  // Use three mutexes to give priority to threads adding/removing nodes
+  // This avoids delays due to quick spinning threads
+  std::mutex add_remove_node_mutex_;
+  std::mutex add_remove_node_high_priority_mutex_;
+  std::mutex add_remove_node_low_priority_mutex_;
```
I'm not a fan of this three-mutex approach, but I think that in general it works ok.

ivanpauno commented on gazebo_ros/src/executor.cpp:
```diff
@@ -58,7 +57,9 @@ void
   rclcpp::node_interfaces::NodeBaseInterface::SharedPtr node_ptr, bool notify)
 {
-  std::lock_guard<std::mutex> wait_lock(wait_mutex_);
+  add_remove_node_high_priority_mutex_.lock();
+  std::lock_guard<std::mutex> lock(add_remove_node_mutex_);
+  add_remove_node_high_priority_mutex_.unlock();
```
```cpp
  std::unique_lock<std::mutex> hplock(add_remove_node_high_priority_mutex_);
  std::lock_guard<std::mutex> lock(add_remove_node_mutex_);
  hplock.unlock();
```

nit: Almost the same, but with RAII.

ivanpauno commented on gazebo_ros/src/executor.cpp:
```diff
@@ -109,31 +112,38 @@ Executor::run()
-    if (!rclcpp::ok(this->context_) || !spinning.load()) {
-      return;
-    }
+    add_remove_node_high_priority_mutex_.lock();
+    std::lock_guard<std::mutex> add_remove_node_lp_lock(add_remove_node_low_priority_mutex_);
```
you have to swap the lines here.
IIUC the idea is that the low priority mutex can only be taken by one low priority thread in the whole critical section.
By taking the hpmutex, then the data mutex, and then unlocking the hpmutex, you notifiy all high priority threads and not the low priority ones (which will then wait in the data mutex).

jacobperron replied:
Oops, yeah you're correct.

ivanpauno commented on gazebo_ros/src/executor.cpp:
```diff
@@ -109,31 +112,38 @@ Executor::run()
-    if (!get_next_executable(any_exec, next_exec_timeout_)) {
-      continue;
-    }
+      if (!get_next_executable(any_exec, std::chrono::nanoseconds(0))) {
```
timeout must be `-1` not `0` to get blocking behavior, which IIUC is the desired behavior here

jacobperron replied:
Right, I meant `-1`. However, by blocking here we hit the same deadlock I was getting before. For example, in the test the following line will block forever because one or more spin threads are retaining a lock on the mutex for adding/removing nodes:

https://github.com/ros-simulation/gazebo_ros_pkgs/blob/f9c0b54c4d8b52a3a1cdd23f242daf53b2e88d1b/gazebo_ros/test/test_executor.cpp#L38

This is why I originally had a timeout; I can't see a better way.

ivanpauno replied:
I see, the thing is that we shouldn't hold that lock while calling `rcl_wait()`, and IIUC we actually don't need to.
We could fix this more cleanly in `rclcpp`, though I think it's acceptable to use a timeout for the moment.

are we planning to fix this in `rclcpp` or is there a reason why we are making the fix here?

jacobperron replied:
> are we planning to fix this in rclcpp or is there a reason why we are making the fix here?

I opted to fix it here instead because I'm not sure the how to fix the issue in rclcpp.

ivanpauno replied:
See https://github.com/ros2/rclcpp/pull/1505.

Could you check if that solves the issue here?

jacobperron commented:
Replaced by https://github.com/ros-simulation/gazebo_ros_pkgs/pull/1212
